File: server/src/main/java/com/lightbend/java/modelServer/ModelServingJob.java
package com.lightbend.java.modelServer;

import com.lightbend.java.modelServer.flinkprocessors.ModelProcessor;
import com.lightbend.java.modelServer.flinkprocessors.RouterProcessor;
import com.lightbend.java.modelServer.flinkprocessors.SpeculativeProcessor;
import com.lightbend.java.modelServer.model.DataConverter;
import com.lightbend.java.modelServer.model.ModelToServe;
import com.lightbend.java.modelServer.model.speculative.SpeculativeModel;
import com.lightbend.java.modelServer.model.speculative.SpeculativeRequest;
import com.lightbend.java.modelServer.model.speculative.SpeculativeServiceRequest;
import com.lightbend.java.modelServer.typeschema.ByteArraySchema;
import com.lightbend.kafka.configuration.java.ModelServingConfiguration;
import com.lightbend.model.Cpudata;
import com.lightbend.speculative.Speculativedescriptor;
import org.apache.flink.api.common.typeinfo.PrimitiveArrayTypeInfo;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.configuration.JobManagerOptions;
import org.apache.flink.configuration.QueryableStateOptions;
import org.apache.flink.configuration.TaskManagerOptions;
import org.apache.flink.runtime.concurrent.Executors;
import org.apache.flink.runtime.highavailability.HighAvailabilityServicesUtils;
import org.apache.flink.runtime.jobgraph.JobGraph;
import org.apache.flink.runtime.minicluster.LocalFlinkMiniCluster;
import org.apache.flink.streaming.api.TimeCharacteristic;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer011;
import org.apache.flink.util.Collector;
import org.apache.flink.util.OutputTag;

import java.util.Optional;
import java.util.Properties;

public class ModelServingJob {
    public static void main(String[] args) {
//    executeLocal();
        executeServer();
    }

    // Execute on the local Flink server - to test queariable state
    private static void  executeServer() {

        // We use a mini cluster here for sake of simplicity, because I don't want
        // to require a Flink installation to run this demo. Everything should be
        // contained in this JAR.

        int port = 6124;
        int parallelism = 2;

        Configuration config = new Configuration();
        config.setInteger(JobManagerOptions.PORT, port);
        config.setString(JobManagerOptions.ADDRESS, "localhost");
        config.setInteger(TaskManagerOptions.NUM_TASK_SLOTS, parallelism);

        // In a non MiniCluster setup queryable state is enabled by default.
        config.setString(QueryableStateOptions.PROXY_PORT_RANGE, "9069");
        config.setInteger(QueryableStateOptions.PROXY_NETWORK_THREADS, 2);
        config.setInteger(QueryableStateOptions.PROXY_ASYNC_QUERY_THREADS, 2);

        config.setString(QueryableStateOptions.SERVER_PORT_RANGE, "9067");
        config.setInteger(QueryableStateOptions.SERVER_NETWORK_THREADS, 2);
        config.setInteger(QueryableStateOptions.SERVER_ASYNC_QUERY_THREADS, 2);

        try {

            // Create a local Flink server
            LocalFlinkMiniCluster flinkCluster = new LocalFlinkMiniCluster(
                    config,
                    HighAvailabilityServicesUtils.createHighAvailabilityServices(
                            config,
                            Executors.directExecutor(),
                            HighAvailabilityServicesUtils.AddressResolution.TRY_ADDRESS_RESOLUTION),
                    false);
            // Start server and create environment
            flinkCluster.start(true);

            StreamExecutionEnvironment env = StreamExecutionEnvironment.createRemoteEnvironment("localhost", port);
            env.setParallelism(parallelism);
            // Build Graph
            buildGraph(env);
            JobGraph jobGraph = env.getStreamGraph().getJobGraph();
            // Submit to the server and wait for completion
            flinkCluster.submitJobAndWait(jobGraph, false);
        } catch (Throwable t){
            t.printStackTrace();
        }
    }

    // Execute locally in the environment
    private static void  executeLocal(){
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        buildGraph(env);
        System.out.println("[info] Job ID: " + env.getStreamGraph().getJobGraph().getJobID());
        try {
            env.execute();
        }
        catch (Throwable t){
            t.printStackTrace();
        }
    }

    // Build execution Graph
    private static void buildGraph(StreamExecutionEnvironment env) {
        env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
        env.enableCheckpointing(5000);

        // Side outputs
        OutputTag<SpeculativeModel> modelTag = new OutputTag<SpeculativeModel>("speculative-model"){};
        OutputTag<SpeculativeRequest> dataTag = new OutputTag<SpeculativeRequest>("speculative-data"){};


        // configure Kafka consumer
        // Data
        Properties dataKafkaProps = new Properties();
        dataKafkaProps.setProperty("zookeeper.connect", ModelServingConfiguration.ZOOKEEPER_HOST);
        dataKafkaProps.setProperty("bootstrap.servers", ModelServingConfiguration.KAFKA_BROKER);
        dataKafkaProps.setProperty("group.id", ModelServingConfiguration.DATA_GROUP);
        // always read the Kafka topic from the current location
        dataKafkaProps.setProperty("auto.offset.reset", "latest");

        // Model
        Properties modelKafkaProps = new Properties();
        modelKafkaProps.setProperty("zookeeper.connect", ModelServingConfiguration.ZOOKEEPER_HOST);
        modelKafkaProps.setProperty("bootstrap.servers", ModelServingConfiguration.KAFKA_BROKER);
        modelKafkaProps.setProperty("group.id", ModelServingConfiguration.MODELS_GROUP);
        // always read the Kafka topic from the beginning
        modelKafkaProps.setProperty("auto.offset.reset", "earliest");

        // Speculative configuration
        Properties speculativeKafkaProps = new Properties();
        speculativeKafkaProps.setProperty("zookeeper.connect", ModelServingConfiguration.ZOOKEEPER_HOST);
        speculativeKafkaProps.setProperty("bootstrap.servers", ModelServingConfiguration.KAFKA_BROKER);
        speculativeKafkaProps.setProperty("group.id", ModelServingConfiguration.SPECULATIVE_GROUP);
        // always read the Kafka topic from the current location
        speculativeKafkaProps.setProperty("auto.offset.reset", "earliest");


        // create a Kafka consumers
        // Data
        FlinkKafkaConsumer011<byte[]> dataConsumer = new FlinkKafkaConsumer011<>(
                ModelServingConfiguration.DATA_TOPIC,
                new ByteArraySchema(),
                dataKafkaProps);

        // Model
        FlinkKafkaConsumer011<byte[]> modelConsumer = new FlinkKafkaConsumer011<>(
                ModelServingConfiguration.MODELS_TOPIC,
                new ByteArraySchema(),
                modelKafkaProps);

        // Speculative
        FlinkKafkaConsumer011<byte[]> speculativeConsumer = new FlinkKafkaConsumer011<>(
                ModelServingConfiguration.SPECULATIVE_TOPIC,
                new ByteArraySchema(),
                speculativeKafkaProps);


        // Create input data streams
        DataStream<byte[]> modelsStream = env.addSource(modelConsumer, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO);
        DataStream<byte[]> dataStream = env.addSource(dataConsumer, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO);
        DataStream<byte[]> speculativeStream = env.addSource(speculativeConsumer, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO);

        // Read data from streams
        DataStream<ModelToServe> models = modelsStream
                .flatMap((byte[] value, Collector<ModelToServe> out) -> {
                    Optional<ModelToServe> model = DataConverter.convertModel(value);
                    if (model.isPresent())
                        out.collect(model.get());
                    else
                        System.out.println("Failed to convert model input");
                }).returns(ModelToServe.class)
                .keyBy(model -> model.getDataType());


        DataStream<Cpudata.CPUData> data = dataStream
                .flatMap((byte[] value, Collector<Cpudata.CPUData> out) -> {
                    Optional<Cpudata.CPUData> record = DataConverter.convertData(value);
                    if (record.isPresent())
                        out.collect(record.get());
                    else
                        System.out.println("Failed to convert data input");
                }).returns(Cpudata.CPUData.class)
                .keyBy(record -> record.getDataType());

        DataStream<Speculativedescriptor.SpeculativeDescriptor> config = speculativeStream
                .flatMap((byte[] value, Collector<Speculativedescriptor.SpeculativeDescriptor> out) -> {
                    Optional<Speculativedescriptor.SpeculativeDescriptor> record = DataConverter.convertConfig(value);
                    if (record.isPresent())
                        out.collect(record.get());
                    else
                        System.out.println("Failed to convert data input");
                }).returns(Speculativedescriptor.SpeculativeDescriptor.class)
                .keyBy(record -> record.getDatatype());


        // Route models and data
        DataStream<SpeculativeServiceRequest> speculativeStart = data
                .connect(models)
                .process(new RouterProcessor());

        // Pickup side inputs and process them
        DataStream<SpeculativeModel> modelUpdate = ((SingleOutputStreamOperator<SpeculativeServiceRequest>) speculativeStart).getSideOutput(modelTag)
                .keyBy(record -> record.getDataModel());

        DataStream<SpeculativeRequest> dataProcess = ((SingleOutputStreamOperator<SpeculativeServiceRequest>) speculativeStart).getSideOutput(dataTag)
                .keyBy(record -> record.getDataModel());

        // Process individual model
        DataStream<SpeculativeServiceRequest> individualresult = dataProcess
                .connect(modelUpdate)
                .process(new ModelProcessor());


        // Run voting
        speculativeStart.union(individualresult)
                .keyBy(record -> record.getDataType())
                .connect(config)
                .process(new SpeculativeProcessor())
                .map(result -> {
                    if(result.isProcessed() && result.getResult().isPresent())
                        System.out.println("Using model " + result.getModel() + ". Calculated result " + result.getResult().get() + ", expected " + result.getSource() + " calculated in " + result.getDuration() + "ms");
                    else System.out.println("No model available - skipping");
                    return result;
                });
    }
}

File: server/src/main/scala/com/lightbend/modelServer/ModelServingJob.scala
/*
 * Copyright (C) 2017  Lightbend
 *
 * This file is part of flink-ModelServing
 *
 * flink-ModelServing is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.lightbend.modelServer


import java.util._

import com.lightbend.kafka.configuration.ModelServingConfiguration
import com.lightbend.modelServer.handlers.BadDataHandler
import com.lightbend.modelServer.model.speculative.{SpeculativeConverter, SpeculativeModel, SpeculativeRequest}
import com.lightbend.modelServer.flinkprocessors.{ModelProcessor, RouterProcessor, SpeculativeProcessor}
import com.lightbend.modelServer.model.{DataRecord, ModelToServe}
import com.lightbend.modelServer.typeschema.ByteArraySchema
import com.lightbend.model.cpudata.CPUData
import com.lightbend.speculative.speculativedescriptor.SpeculativeDescriptor
import org.apache.flink.api.scala._
import org.apache.flink.configuration._
import org.apache.flink.runtime.concurrent.Executors
import org.apache.flink.runtime.highavailability.HighAvailabilityServicesUtils
import org.apache.flink.runtime.minicluster.LocalFlinkMiniCluster
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala.{OutputTag, StreamExecutionEnvironment}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer011


/**
  * Created by boris on 5/9/17.
  * loosely based on http://dataartisans.github.io/flink-training/exercises/eventTimeJoin.html approach
  * for queriable state
  *   https://github.com/dataArtisans/flink-queryable_state_demo/blob/master/README.md
  * Using Flink min server to enable Queryable data access
  *   see https://github.com/dataArtisans/flink-queryable_state_demo/blob/master/src/main/java/com/dataartisans/queryablestatedemo/EventCountJob.java
  *
  * This little application is based on a RichCoProcessFunction which works on a keyed streams. It is applicable
  * when a single applications serves multiple different models for different data types. Every model is keyed with
  * the type of data what it is designed for. Same key should be present in the data, if it wants to use a specific
  * model.
  * Scaling of the application is based on the data type - for every key there is a separate instance of the
  * RichCoProcessFunction dedicated to this type. All messages of the same type are processed by the same instance
  * of RichCoProcessFunction
  */
object ModelServingJob {

  def main(args: Array[String]): Unit = {
//    executeLocal()
    executeServer()
  }

  // Execute on the local Flink server - to test queariable state
  def executeServer() : Unit = {

    // We use a mini cluster here for sake of simplicity, because I don't want
    // to require a Flink installation to run this demo. Everything should be
    // contained in this JAR.

    val port = 6124
    val parallelism = 2

    val config = new Configuration()
    config.setInteger(JobManagerOptions.PORT, port)
    config.setString(JobManagerOptions.ADDRESS, "localhost")

    // In a non MiniCluster setup queryable state is enabled by default.
    config.setString(QueryableStateOptions.PROXY_PORT_RANGE, "9069")
    config.setInteger(QueryableStateOptions.PROXY_NETWORK_THREADS, 2)
    config.setInteger(QueryableStateOptions.PROXY_ASYNC_QUERY_THREADS, 2)

    config.setString(QueryableStateOptions.SERVER_PORT_RANGE, "9067")
    config.setInteger(QueryableStateOptions.SERVER_NETWORK_THREADS, 2)
    config.setInteger(QueryableStateOptions.SERVER_ASYNC_QUERY_THREADS, 2)


    // Create a local Flink server
    val flinkCluster = new LocalFlinkMiniCluster(
      config,
      HighAvailabilityServicesUtils.createHighAvailabilityServices(
        config,
        Executors.directExecutor(),
        HighAvailabilityServicesUtils.AddressResolution.TRY_ADDRESS_RESOLUTION),
      false)
    try {
      // Start server and create environment
      flinkCluster.start(true)
      val env = StreamExecutionEnvironment.createRemoteEnvironment("localhost", flinkCluster.getLeaderRPCPort)
       // Build Graph
      buildGraph(env)
      val jobGraph = env.getStreamGraph.getJobGraph
      // Submit to the server and wait for completion
      flinkCluster.submitJobAndWait(jobGraph, false)
    } catch {
      case e: Exception => e.printStackTrace()
    }
  }

  // Execute localle in the environment
  def executeLocal() : Unit = {
    val env = StreamExecutionEnvironment.getExecutionEnvironment
    buildGraph(env)
    System.out.println("[info] Job ID: " + env.getStreamGraph.getJobGraph.getJobID)
    env.execute()
  }

  // Build execution Graph
  def buildGraph(env : StreamExecutionEnvironment) : Unit = {
    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
    env.enableCheckpointing(5000)

    // Side outputs
    val modelTag = OutputTag[SpeculativeModel]("speculative-model")
    val dataTag = OutputTag[SpeculativeRequest]("speculative-data")


    import ModelServingConfiguration._

    // configure Kafka consumer

    // Data
    val dataKafkaProps = new Properties
    dataKafkaProps.setProperty("zookeeper.connect", ZOOKEEPER_HOST)
    dataKafkaProps.setProperty("bootstrap.servers", KAFKA_BROKER)
    dataKafkaProps.setProperty("group.id", DATA_GROUP)
    // always read the Kafka topic from the current location
    dataKafkaProps.setProperty("auto.offset.reset", "latest")

    // Model
    val modelKafkaProps = new Properties
    modelKafkaProps.setProperty("zookeeper.connect", ZOOKEEPER_HOST)
    modelKafkaProps.setProperty("bootstrap.servers", KAFKA_BROKER)
    modelKafkaProps.setProperty("group.id", MODELS_GROUP)
    // always read the Kafka topic from the current location
    modelKafkaProps.setProperty("auto.offset.reset", "earliest")

    // Speculative configuration
    val speculativeKafkaProps = new Properties
    speculativeKafkaProps.setProperty("zookeeper.connect", ZOOKEEPER_HOST)
    speculativeKafkaProps.setProperty("bootstrap.servers", KAFKA_BROKER)
    speculativeKafkaProps.setProperty("group.id", SPECULATIVE_GROUP)
    // always read the Kafka topic from the current location
    speculativeKafkaProps.setProperty("auto.offset.reset", "earliest")

    // create a Kafka consumers

    // Data
    val dataConsumer = new FlinkKafkaConsumer011[Array[Byte]](
      DATA_TOPIC,
      new ByteArraySchema,
      dataKafkaProps
    )

    // Model
    val modelConsumer = new FlinkKafkaConsumer011[Array[Byte]](
      MODELS_TOPIC,
      new ByteArraySchema,
      modelKafkaProps
    )

    // Speculative
    val speculativeConsumer = new FlinkKafkaConsumer011[Array[Byte]](
      SPECULATIVE_TOPIC,
      new ByteArraySchema,
      speculativeKafkaProps
    )

    // Create input data streams
    val dataStream = env.addSource(dataConsumer)
    val modelsStream = env.addSource(modelConsumer)
    val speculativeStream = env.addSource(speculativeConsumer)

    // Read data from streams
    val models = modelsStream.map(ModelToServe.fromByteArray(_))
      .flatMap(BadDataHandler[ModelToServe])
      .keyBy(_.dataType)
    val data = dataStream.map(DataRecord.fromByteArray(_))
      .flatMap(BadDataHandler[CPUData])
      .keyBy(_.dataType)
    val config = speculativeStream.map(SpeculativeConverter.fromByteArray(_))
      .flatMap(BadDataHandler[SpeculativeDescriptor])
      .keyBy(_.datatype)

    // Route models and data
    val speculativeStart = data
      .connect(models)
      .process(RouterProcessor())

    // Pickup side inputs and process them
    val modelUpdate = speculativeStart.getSideOutput(modelTag).keyBy(_.dataModel)
    val dataProcess = speculativeStart.getSideOutput(dataTag).keyBy(_.dataModel)

    // Process individual model
    val individualresult = dataProcess
      .connect(modelUpdate)
      .process(ModelProcessor())

    // Run voting
    speculativeStart.union(individualresult).keyBy(_.dataType)
      .connect(config)
      .process(SpeculativeProcessor())
      .map(result => {
        result.processed match {
          case true =>
            result.result match{
              case Some(r) => println(s"Using model ${result.model}. Calculated result $r, expected ${result.source} calculated in ${result.duration} ms")
              case _ =>
            }
          case _ => println("No model available - skipping")
        }
      })
  }
}